backend/cfg/vectorize.ml

```diff
@@ -2,13 +2,145 @@
 
 module DLL = Flambda_backend_utils.Doubly_linked_list
 
-let dump ppf cfg_with_layout ~msg =
+module Instruction = struct
+  module Id = struct
+    include Numbers.Int
+  end
+
+  type t =
+    [ `Basic of Cfg.basic Cfg.instruction
+    | `Terminator of Cfg.terminator Cfg.instruction ]
+
+  let id (instruction : t) : Id.t =
+    match instruction with
+    | `Basic instruction -> instruction.id
+    | `Terminator instruction -> instruction.id
+
+  let arg (instruction : t) : Reg.t Array.t =
+    match instruction with
+    | `Basic instruction -> instruction.arg
+    | `Terminator instruction -> instruction.arg
+
+  let res (instruction : t) : Reg.t Array.t =
+    match instruction with
+    | `Basic instruction -> instruction.res
+    | `Terminator instruction -> instruction.res
+
+  let destroyed (instruction : t) : Reg.t Array.t =
+    match instruction with
+    | `Basic instruction -> Proc.destroyed_at_basic instruction.desc
+    | `Terminator instruction -> Proc.destroyed_at_terminator instruction.desc
+end
+
+module Dependency_graph = struct
+  (* The dependency graph shows dependencies between instructions within the
+     same basic block *)
+  module Node = struct
+    type t =
+      { id : Instruction.Id.t;
+        out_edges : Instruction.Id.Set.t;
+        in_edges : Instruction.Id.Set.t
+      }
+
+    let init id : t =
+      { id;
+        out_edges = Instruction.Id.Set.empty;
+        in_edges = Instruction.Id.Set.empty
+      }
+  end
+
+  type t = Node.t Instruction.Id.Tbl.t
+
+  let init () : t = Instruction.Id.Tbl.create 100
+
+  let from_basic_block (block : Cfg.basic_block) =
+    let dependency_graph = init () in
+    let is_changed_in instruction reg =
+      Array.exists (Reg.same reg) (Instruction.res instruction)
+      || Array.exists (Reg.same reg) (Instruction.destroyed instruction)
+    in
+    let latest_change ~(current : Instruction.Id.t) (reg : Reg.t) =
+      DLL.fold_right block.body
```

Comment:
This loop would be more natural and slightly faster using `DLL.last` and `DLL.prev`.

Reply:
you mean with a while loop?

Reply:
I meant `fold_right`.